src/MetricExporter.ts

```diff
@@ -82,49 +90,132 @@ export class DynatraceMetricExporter implements MetricExporter {
+			});
+		}
+
+		const dtMetrics = metrics.map((metric) => {
+			const dimensions = Object.entries(metric.labels).map(([key, value]) => ({ key, value }));
+			switch (metric.aggregator.kind) {
+				case AggregatorKind.SUM: {
+					const data = metric.aggregator.toPoint();
+					const normalizedMetric = this._dtMetricFactory
+						.createTotalCounter(
+							metric.descriptor.name,
+							dimensions,
+							data.value,
+							hrTimeToMilliseconds(data.timestamp)
+						);
+					if (normalizedMetric == null) {
+						this._warnNormalizationFailure(metric.descriptor.name);
+					}
+					return normalizedMetric;
+				}
+				case AggregatorKind.HISTOGRAM: {
+					const data = metric.aggregator.toPoint();
+					const { sum, count } = data.value;
+
+					// TODO remove this hack which pretends all data points had the same value
+					const avg = sum / count;
 
-		if (!payload) {
+					const value: SummaryValue = {
+						min: avg,
+						max: avg,
+						sum,
+						count
+					};
+					const normalizedMetric = this._dtMetricFactory
+						.createSummary(
+							metric.descriptor.name,
+							dimensions,
+							value,
+							hrTimeToMilliseconds(data.timestamp)
+						);
+					if (normalizedMetric == null) {
+						this._warnNormalizationFailure(metric.descriptor.name);
+					}
+					return normalizedMetric;
+				}
+				case AggregatorKind.LAST_VALUE: {
+					const data = metric.aggregator.toPoint();
+					const normalizedMetric = this._dtMetricFactory
+						.createGauge(
+							metric.descriptor.name,
+							dimensions,
+							data.value,
+							hrTimeToMilliseconds(data.timestamp)
+						);
+					if (normalizedMetric == null) {
+						this._warnNormalizationFailure(metric.descriptor.name);
+					}
+					return normalizedMetric;
+				}
+			}
+		});
+
+		const lines = dtMetrics
+			.filter((m): m is Metric => m != null)
+			.map(m => m.serialize());
+
+		if (lines.length === 0) {
 			process.nextTick(resultCallback, { code: ExportResultCode.SUCCESS });
 			return;
 		}
 
+		const payload = lines.join("\n");
+
 		const request = this._httpRequest(this._reqOpts);
 		const self = this;
 
 		function onResponse(res: http.IncomingMessage) {
-			self._logger.debug(`request#onResponse: statusCode: ${res.statusCode}`);
+			api.diag.debug(`request#onResponse: statusCode: ${res.statusCode}`);
 
-			res.resume(); // discard any incoming data
 
 			res.on("error", e => {
 				// no need for handling a response error as a valid statusCode has
 				// been received before which indicates message was received
-				self._logger.debug(`response#error: ${e.message}`);
+				api.diag.debug(`response#error: ${e.message}`);
 			});
 
 			if (
 				res.statusCode != null &&
 				res.statusCode >= 200 &&
 				res.statusCode < 300
 			) {
+				res.resume(); // discard any incoming data
 				process.nextTick(resultCallback, { code: ExportResultCode.SUCCESS });
 			} else if (res.statusCode === 401 || res.statusCode === 403) {
```

Comment:
There is also 400 if only some data points were valid but some some discarded:
https://www.dynatrace.com/support/help/shortlink/api-metrics-v2-post-datapoints#response-codes
For the discarded ones this would be a permanent error since it's a formatting or sanity check error.

Reply:
In the case of 400 it would fall to the else block. There would be a warn level log for the status code and a debug level log which would show the failure reponse from dynatrace showing some logs were discarded. Is that enough or should I make a change?

Reply:
Alright then let's add a comment to make clear that this else block does not only cover unexpected cases but also handles the well-expected 400 properly, e.g.
```
} else { // e.g., 400 if some lines were valid while some violated the restrictions
```